Modify rule S6214: Convert to LaYC format (#2633)
SonarSource/rspec: rules/S6214/cfamily/rule.adoc

> Compare integers safely using `std::cmp_*` functions to avoid any unexpected results.

Reviewer: ```suggestion
Compare integers of mixed signedness safely using `std::cmp_*` functions to avoid any unexpected results.
```

== Why is this an issue?

Comparison between `signed` and `unsigned` integers is dangerous because it produces counter-intuitive results due to implicit conversions.
When a signed integer is compared to an unsigned one, the former might be converted to unsigned.

> The conversion preserves the two's-complement bit pattern of the signed value that often corresponds to a large unsigned result.

Reviewer: Is this guaranteed to be two's-compliment?

Author: Yes, part of the cpp20 standard, and the rule is aimed "since-cpp20"

Reviewer: To avoid confusing the reader, I would qualify this statement with "Since C++20, ...". No need to specify what happens in other standards since the rule is applicable only starting from C++20

For example, `2U < -1` is `true`.

This rule raises an issue when an unsigned integer is compared with a negative value.

=== What is the potential impact

> Integer comparisons are usually used in creating branch conditions.

Reviewer: ```suggestion
Integer comparisons are often used in branch conditions.
```

An unexpected result from one of these conditions can lead to hard-to-detect issues.

For example, using container size functions in a comparison can lead to such a problem since these return an unsigned integer value.

== How to fix it

{cpp}20 introduced a remedy to this common pitfall: a family of `std::cmp_*` functions defined in the `<utility>` header.
These functions correctly handle negative numbers and lossy integer conversion.
For example, `std::cmp_less(2U, -1)` is `false`.

=== Code Examples

==== Noncompliant code example

> [source,cpp]

Reviewer: I think the diff view here will be helpful

Compare integers of mixed signedness safely using `std::cmp_*` functions to avoid any unexpected results.

== Why is this an issue?

Comparison between `signed` and `unsigned` integers is dangerous because it produces counter-intuitive results due to implicit conversions.
When a signed integer is compared to an unsigned one, the former might be converted to unsigned.

> Since {cpp}20, the conversion preserves the two's-complement bit pattern of the signed value that often corresponds to a large unsigned result.

Reviewer: Depending on your definition of large, but given the sign bit, the negative number converted to unsigned will occupy the largest half of the unsigned range, i.e., will always be bigger than any number representable in the signed type of the same width, and that sounds pretty large to me
```suggestion
Since {cpp}20, the conversion preserves the two's-complement bit pattern of the signed value that always corresponds to a large unsigned result.
```

> ----

Reviewer: In the compliant solution you have two more examples:
```
bool fun(int x, std::vector<int> const& v) {
  return std::cmp_less(x, v.size()); // Compliant
}
std::vector<int> v = foo();
if (0 < v.size() && v.size() < 100) { // Compliant, even though v.size() returns an unsigned integer
}
```
It would be instructive to also show their noncompliant version

== Resources

> 

Reviewer: ```suggestion

=== Related rules

```

* S845 - a more generic rule about mixing signed and unsigned values.

> * S6183 - a version of this rule that triggers when signed and unsigned variables are compared.

Reviewer: ```suggestion
* S6183 - a version of this rule that triggers when any signed and unsigned variables are compared.
```

}

std::vector<int> v = foo();
if (0 <= v.size() && v.size() < 100) { // Compliant, even though v.size() returns an unsigned integer
}
----

> == Resources

Reviewer: Perhaps you can also point to the documentation of the cmp_* functions https://en.cppreference.com/w/cpp/utility/intcmp
